src/pip/_internal/utils/pyc_compile.py

```diff
@@ -0,0 +1,170 @@
+# -------------------------------------------------------------------------- #
+# NOTE: Importing from pip's internals or vendored modules should be AVOIDED
+#       so this module remains fast to import, minimizing the overhead of
+#       spawning a new bytecode compiler worker.
+# -------------------------------------------------------------------------- #
+
+import compileall
+import importlib
+import os
+import sys
+import warnings
+from collections.abc import Callable, Iterable, Iterator
+from contextlib import contextmanager, redirect_stdout
+from io import StringIO
+from pathlib import Path
+from typing import TYPE_CHECKING, Literal, NamedTuple, Optional, Protocol, Union
+
+if TYPE_CHECKING:
+    from pip._vendor.typing_extensions import Self
+
+WorkerSetting = Union[int, Literal["auto"]]
+
+CODE_SIZE_THRESHOLD = 1000 * 1000  # 1 MB of .py code
+WORKER_LIMIT = 8
+
+
+@contextmanager
+def _patch_main_module_hack() -> Iterator[None]:
+    """Temporarily replace __main__ to reduce the worker startup overhead.
+
+    concurrent.futures imports the main module while initializing new workers
+    so any global state is retained in the workers. Unfortunately, when pip
+    is run from a console script wrapper, the wrapper unconditionally imports
+    pip._internal.cli.main and everything else it requires. This is *slow*.
+
+    The compilation code does not depend on any global state, thus the costly
+    re-import of pip can be avoided by replacing __main__ with any random
+    module that does nothing.
+    """
```

Comment:
This is ugly, but this does reduce worker startup overhead significantly. I couldn't measure an improvement on Windows, but Linux and macOS both saw major drops in startup overhead. See also: https://github.com/pypa/pip/issues/12712#issuecomment-2660731338.